### PR TITLE
feat(performance): Add support for team key transactions in end points

### PR DESCRIPTION
**Pull request by Zylphrex**
This adds in support for team-based key transactions in the key transactions end
points behind the `team-key-transactions` flag.

### CODE REVIEW

**Review by k-fish — APPROVED** [review]
Looks like the methods are all behind a feature flag, so 👏 . I'm not as familiar but I'm assuming none of these filters or bulk_create will have scaling issues on prod (eg. have the right indexes etc)?

File: src/sentry/discover/endpoints/discover_key_transactions.py [hunk 1]
```diff
@@ -43,51 +45,143 @@ def get(self, request, organization):
...
+                if serializer.is_valid():
+                    data = serializer.validated_data
+                    base_filter["transaction"] = data["transaction"]
+                    base_filter["project"] = project
+
+                    if KeyTransaction.objects.filter(**base_filter).exists():
+                        return Response(status=204)
+
+                    try:
+                        KeyTransaction.objects.create(**base_filter)
```
Comment (shruthilayaj) [inline src/sentry/discover/endpoints/discover_key_transactions.py:90]: What about using a KeyTransaction.objects.get_or_create pattern here?

File: src/sentry/discover/endpoints/discover_key_transactions.py [hunk 2]
```diff
@@ -43,51 +45,143 @@ def get(self, request, organization):
...
-                if KeyTransaction.objects.filter(**base_filter).exists():
+                base_filter = {
+                    "organization": organization,
+                    "project": project,
+                    "transaction": data["transaction"],
+                }
+
+                keyed_transactions = TeamKeyTransaction.objects.filter(
+                    **base_filter, team__in=data["team"]
+                )
```
Comment (shruthilayaj) [inline src/sentry/discover/endpoints/discover_key_transactions.py:113]: I think you can just do `.values_list("team_id", flat=True)` here?

File: src/sentry/discover/endpoints/discover_key_transactions.py [hunk 3]
```diff
@@ -43,51 +45,143 @@ def get(self, request, organization):
...
+                "organization": organization,
+                "project": project,
+                "transaction": data["transaction"],
+            }
+
+            try:
+                model = TeamKeyTransaction.objects.filter(
+                    **base_filter, team__in=data["team"]
+                ).delete()
+            except TeamKeyTransaction.DoesNotExist:
```
Comment (shruthilayaj) [inline src/sentry/discover/endpoints/discover_key_transactions.py:174]: I don't think `filter(...).delete()` would ever raise a DoesNotExist, so you can probably skip the try/except

```
In [5]: TeamKeyTransaction.objects.filter(team=team)
Out[5]: <BaseQuerySet []>

In [6]: TeamKeyTransaction.objects.filter(team=team).delete()
Out[6]: (0, {'sentry.TeamKeyTransaction': 0})
```

File: src/sentry/discover/endpoints/serializers.py [hunk 4]
```diff
@@ -249,3 +256,37 @@ def validate(self, data):
...
+                    f"You do not have permission to access {team.name}"
+                )
+
+        return teams
+
+    def validate(self, data):
+        data = super().validate(data)
+        if self.context.get("mode") == "create":
+            team = data["team"]
+            count = TeamKeyTransaction.objects.filter(team_id__in=[t.id for t in team]).count()
```
Comment (shruthilayaj) [inline src/sentry/discover/endpoints/serializers.py:286]: This limit will be applied per team

File: src/sentry/discover/endpoints/discover_key_transactions.py (hunk 1 above)
Comment (Zylphrex) [reply src/sentry/discover/endpoints/discover_key_transactions.py:90]: This is the existing `KeyTransaction`, going to leave that unchanged (at least in this PR).

File: src/sentry/discover/endpoints/serializers.py (hunk 4 above)
Comment (Zylphrex) [reply src/sentry/discover/endpoints/serializers.py]: fixed in b9be6be

File: src/sentry/discover/endpoints/discover_key_transactions.py (hunk 3 above)
Comment (Zylphrex) [reply src/sentry/discover/endpoints/discover_key_transactions.py]: removed in b9be6be

File: src/sentry/discover/endpoints/discover_key_transactions.py (hunk 2 above)
Comment (Zylphrex) [reply src/sentry/discover/endpoints/discover_key_transactions.py:115]: fixed in b9be6be

**Review by shruthilayaj — APPROVED** [review]
LGTM

**Pull request merged**